### PR TITLE
Update packages, configs + Node.js 20

### PR DESCRIPTION
**Pull request by colinrotherham**
Similar to https://github.com/alphagov/accessible-autocomplete/pull/316 this PR brings all our dependencies up to date

Excluding two major bumps which I've split out:

1. Upgrade to [webpack v5](https://webpack.js.org/blog/2020-10-10-webpack-5-release/) in #614
2. Upgrade to [WebdriverIO v8](https://webdriver.io/blog/2022/12/01/webdriverio-v8-released/) in #615

**Note:** Updates to Preact have also been excluded since we'll need to update tests separately. For example, we currently [require a mismatched Preact version](https://github.com/alphagov/accessible-autocomplete/blob/92a1c7200eba68e8f8e8f7d0926bb1fabaaf7397/examples/preact/index.html#L63-L64) via https://unpkg.com

## Configs, tests and polyfills

We know the [deprecated `@wdio/sync` package won't install](https://github.com/alphagov/accessible-autocomplete/issues/608) so tests now use async/await instead

But we can remove more (see below) with config updates, especially now [IE 8–10 support is clarified](https://github.com/alphagov/design-system-team-internal/issues/561):

1. Browserslist config updated with IE11+ to match 
2. Babel config presets updated for [`core-js@3`](https://www.npmjs.com/package/core-js) polyfills
4. Babel config transforms for ES3 removed (prevented errors in IE8)
5. Node.js version config **.nvmrc** updated to v20.9.0 `lts/iron`
6. Husky setup moved to new `.husky/pre-commit` config format

## Packages removed

The following packages were unnecessary or unused:

```console
@wdio/sync
copy-webpack-plugin
dotenv
karma-chai
karma-chai-sinon
replace-bundle-webpack-plugin
sinon
sinon-chai
```

With more included in [`@babel/preset-env`](https://www.npmjs.com/package/@babel/preset-env) automatically now or to transform to ES3 for Internet Explorer 8:

```console
@babel/plugin-proposal-class-properties
@babel/plugin-proposal-decorators
@babel/plugin-transform-member-expression-literals
@babel/plugin-transform-modules-commonjs
@babel/plugin-transform-property-literals
```

## Package audit

See results from `npm audit` below

### Before

```console
found 181 vulnerabilities (3 low, 45 moderate, 104 high, 29 critical) in 2141 scanned packages
  run `npm audit fix` to fix 132 of them.
  43 vulnerabilities require semver-major dependency updates.
  6 vulnerabilities require manual review. See the full report for details.
```

### After

```console
11 vulnerabilities (3 moderate, 8 high)

To address all issues (including breaking changes), run:
  npm audit fix --force
```

See PRs #614 and #615 to reach `0 vulnerabilities`

### CODE REVIEW

**Comment by rgarner** [comment]
@colinrotherham  Wow, thanks, this is a lot of work and looks great. The move to `await` (away from fibers?) has fixed my `Promise` issues with integration tests — `npm test` now completes with no warnings or errors. I had one issue on setup — `npm install` (from npm `6.14.18`) downgraded all `package.json` `integrity` checksums from `sha-512` to `sha1`, leaving me with a modified `package-lock.json` in my working tree. For anyone else upgrading with a similar issue (maybe [this one?](https://github.com/npm/cli/issues/2401)), here's what worked for me:

```bash
git checkout package-lock.json
rm -rf node_modules
npm cache clear --force # you may not need this, at your own risk, etc — I did need it, though
npm install
```

Now rebasing #611 to see how that goes (to add an integration test)...

**Comment by colinrotherham** [comment]
Brilliant, thanks for confirming that

>…I had one issue on setup — `npm install` (from npm `6.14.18`) downgraded all `package.json` `integrity` checksums from `sha-512` to `sha1`, leaving me with a modified `package-lock.json` in my working tree…

Sounds like we might be on different npm versions?

```console
$ node -v
v14.21.3
```

```console
$ npm -v
6.14.18
```

Hope this idea is alright, but I'll cherry your **.nvmrc** commit into my branch and we'll close this PR

Gives us a single place to rebase from and I can work on raising the maximum Node.js version too

**Comment by colinrotherham** [comment]
Latest push highlights includes:

* Update to Node.js 20
* Update to [WebdriverIO v8](https://webdriver.io/blog/2022/12/01/webdriverio-v8-released/)
* Update to Puppeteer v21

**Comment by rgarner** [comment]
@colinrotherham this is all working perfectly for me over on this branch, can run every type of test, so whatever I've wrought upon #611 is of my own rebase making and I shall unpick that in the morning!

File: package.json [hunk 1]
```diff
@@ -18,86 +18,69 @@
...
     "karma-mocha-reporter": "^2.2.5",
-    "karma-sourcemap-loader": "^0.3.7",
+    "karma-sourcemap-loader": "^0.4.0",
     "karma-webpack": "^4.0.2",
-    "mocha": "^5.2.0",
+    "mocha": "^10.2.0",
     "npm-run-all": "^4.1.5",
-    "puppeteer": "^13.3.1",
-    "replace-bundle-webpack-plugin": "^1.0.0",
-    "sinon": "^6.3.5",
```
Comment (rgarner) [inline package.json:80]: Ah, I was [using](https://github.com/alphagov/accessible-autocomplete/pull/611/commits/84909c65471014b3189f29a33c19101f8429c1af#diff-4a14e2d8854c9e6cdf908f096dc93ec20897e9cc747b1515125ecfe0ad38bfa8R241) `useFakeTimers` from sinon to test the debounce.

Comment (rgarner) [reply package.json:80]: (I can do that with `setTimeout` and `done()` at the cost of some milliseconds)

Comment (colinrotherham) [reply package.json:80]: Thanks @rgarner, yeah probably best to use real timers in browsers

**Comment by colinrotherham** [comment]
Thanks @ryanb-gds, we've been watching https://github.com/alphagov/accessible-autocomplete-multiselect/pull/45 👀

Good to see that work happening

> Hello folks, is there anything I can do to help get this merged? My team maintains [a fork of this repo](https://github.com/alphagov/accessible-autocomplete-multiselect) which needs upgrading to a more recent node version, would love to rebase it onto this if this is close to merging.

It's in our "Needs review" column (@romaricpascal) but will likely be picked up after GOV.UK Frontend v5

We'd welcome code reviews if you could help spot any issues?

### Reducing browser support
One tricky change in this PR would be https://github.com/alphagov/accessible-autocomplete/pull/612/commits/8b77b5f7fe07b22ffb03e4bd2113fad29bf71990 where polyfills and transforms for unsupported browsers were unintentionally left in the `accessible-autocomplete@2.0.0` release and have been removed

Although we no longer [support or test in Internet Explorer 9–10](https://github.com/alphagov/design-system-team-internal/issues/561) and will [drop IE11 JavaScript support soon](https://github.com/alphagov/govuk-frontend/discussions/2607) we may want to defer this commit until `accessible-autocomplete@3.0.0`

**Comment by ryanb-gds** [comment]
Thanks @colinrotherham, I would be more than happy to do a review. I will take a look today if I can.

I'm not quite sure about the future of the multiselect at the moment. We've struggled to make it as accessible as we'd like, but more accessible design alternatives feel less practical for many of our users (who are civil servants rather than the public). We may explore some more options though. Let me know if there are any related conversations you'd be interested in having.

**Comment by ryanb-gds** [comment]
It doesn't look like I did 🤔 not sure where that went. The only minor suggestion was that in the integration test some assertions that await two conditions (e.g. [here](https://github.com/alphagov/accessible-autocomplete/blob/b3353884d76471c163fb0c6743b745a50bb0e369/test/integration/index.js#L58)) could perhaps use `Promise.all`, but I suspect you've intentionally chosen to stay as close to the original code as possible.

**Comment by colinrotherham** [comment]
Looks like the [`chai@5` ES modules update](https://github.com/chaijs/chai/releases/tag/v5.0.0) will need some work so I'll roll it back

We can let Dependabot pick it up another day via:

* https://github.com/alphagov/accessible-autocomplete/pull/613

**Review by romaricpascal — APPROVED** [review]
Great to have everything a little more up to date, thanks for going through this (and having everything nicely split for review as well). ⛵

**Pull request merged**